File: src/Python/core/registry.py
from __future__ import annotations
from typing import Dict, Optional, Callable, TypeVar, Generic

T = TypeVar('T')

class PipelineRegistry(Generic[T]):
    def __init__(self):
        self._map: Dict[str, Callable[[T], T]] = {}

    def register(self, key: str, pipeline: Callable[[T], T]) -> None:
        self._map[key] = pipeline

    def lookup(self, key: str) -> Optional[Callable[[T], T]]:
        return self._map.get(key)

    def as_map(self):
        return dict(self._map)

    def size(self) -> int:
        return len(self._map)

File: src/typescript/core/registry.ts
export class PipelineRegistry<T> {
  private mapping: Map<string, (value: T) => T | Promise<T>>;

  constructor() {
    this.mapping = new Map();
  }

  register(key: string, pipelineCallable: (value: T) => T | Promise<T>): void {
    this.mapping.set(key, pipelineCallable);
  }

  lookup(key: string): ((value: T) => T | Promise<T>) | undefined {
    return this.mapping.get(key);
  }

  asMap(): Map<string, (value: T) => T | Promise<T>> {
    return new Map(this.mapping);
  }

  size(): number {
    return this.mapping.size;
  }
}
